scipy/stats/tests/test_morestats.py

```diff
@@ -522,6 +522,156 @@ def test_result_attributes(self):
         attributes = ('statistic', 'pvalue')
         check_named_results(res, attributes)
 
+    def test_bad_alternative(self):
+        # invalid value for alternative must raise a ValueError
+        x1 = [1, 2, 3, 4]
+        x2 = [5, 6, 7, 8]
+        assert_raises(ValueError, stats.ansari, x1, x2, alternative='foo')
+
+    def test_bad_mode(self):
+        # invalid value for mode must raise a ValueError
+        x1 = [1, 2, 3, 4]
+        x2 = [5, 6, 7, 8]
+        assert_raises(ValueError, stats.ansari, x1, x2, mode='foo')
+
+    def test_alternative_exact(self):
+        x1 = [-5, 1, 5, 10, 15, 20, 25] # high scale, loc=10
+        x2 = [7.5, 8.5, 9.5, 10.5, 11.5, 12.5] # low scale, loc=10
+        # ratio of scales is greater than 1. So, the
+        # p-value must be high when `alternative='less'`
+        # and low when `alternative='greater'`.
+        pval_l = stats.ansari(x1, x2, alternative='less').pvalue
+        pval_g = stats.ansari(x1, x2, alternative='greater').pvalue
+        assert pval_l > 0.95
+        assert pval_g < 0.05 # level of significance.
+        # sanity check. The result should flip if
+        # we exchange x and y.
+        pval_l_reverse = stats.ansari(x2, x1, alternative='less').pvalue
+        pval_g_reverse = stats.ansari(x2, x1, alternative='greater').pvalue
+        assert pval_l_reverse < 0.05
+        assert pval_g_reverse > 0.95
```

Comment:
One check that would make sense to add here is that the sum of the two values - or maybe the sum of the two p-values should equal one plus (or minus, depending on the answer to an earlier comment) the probability mass _at_ the observed value of the test statistic. Hmm but I see that the code is not structured to easily get at that. Would have been nice if the code for the distribution of the test statistic had been broken out as a class. Let's not refactor that here, though. Just musings, I guess.

Reply:
> One check that would make sense to add here is that the sum of the two values - or maybe the sum of the two p-values should equal one plus (or minus, depending on the answer to an earlier comment) the probability mass at the observed value of the test statistic. Hmm but I see that the code is not structured to easily get at that. Would have been nice if the code for the distribution of the test statistic had been broken out as a class. Let's not refactor that here, though. Just musings, I guess.

Yeah, I think we can do that too. We can use `scipy.stats.statlib.gscale` to generate the distribution and use it to test if the sum of p-values equal one plus the probability mass at the observed value of the test statistic. Let me do that and then we can decide whether we want to keep it or not.